public/app/plugins/datasource/graphite/datasource.ts

```diff
@@ -473,23 +481,30 @@ export class GraphiteDatasource extends DataSourceApi<GraphiteQuery, GraphiteOpt
       httpOptions.params.until = this.translateTime(options.range.to, true, options.timezone);
     }
 
-    return this.doGraphiteRequest(httpOptions).then((results: any) => {
-      if (results.data) {
-        return _.map(results.data, (tag) => {
-          return { text: tag };
-        });
-      } else {
+    return this.doGraphiteRequest(httpOptions)
+      .then((results: any) => {
+        if (results.data) {
+          return _.map(results.data, (tag) => {
+            return { text: tag };
+          });
+        } else {
+          return [];
+        }
+      })
+      .catch((error) => {
+        console.error(error);
+        this.displayTagsAutocompleteAlert(error, url);
```

Comment:
I don't think this is a good place to show the error. Error handling should be in general pushed as high in stack as possible to the place where it is clear what to do about it. Datasource code could be possible used from multiple contexts like alerting or other datasources (none of it is true right now but could be in the future) so it can't really know if showing an user error is appropriate. Best way to handle this would be to throw this error and let the caller handle it. In this case I assume it would be the queryEditor component where it is clear that user is interacting with the autocomplete and so makes sense to show him an alert notification.

Reply:
Good point! I've moved error handling to `query_ctrl.ts`. I've also converted `appEvents.emit` to `notifyApp` in `checkOtherSegments()`.